### PR TITLE
Support aggregate event IDs for multi-toolkit MCP proxy

### PR DESCRIPTION
**Pull request by jfallows**
## Description

Implements support for aggregate event IDs in the MCP proxy binding to enable proper event tracking and resumption across multiple toolkit routes.

### Changes

**Core Implementation:**
- Added `McpAggregateEventId` utility class to encode/decode composite event IDs using toolkit-specific prefixes (e.g., `"S=100;2=200"`)
- Prefixes are computed from toolkit names using CRC32C hashing to ensure uniqueness up to 6 characters
- Supports encoding multiple per-route event IDs into a single aggregate ID and decoding them back for route-specific resumption

**Lifecycle Management:**
- Enhanced `McpLifecycleServer` to track per-route event IDs via `lastEventIdsByPrefix` array
- Added `aggregating()` check to determine if a session spans multiple routes
- Implemented `recordRouteEventId()` to update tracked IDs from individual routes
- Implemented `mintAggregateEventId()` to generate composite IDs for client notification
- Implemented `dispatchAggregateResume()` to decompose aggregate IDs and resume individual routes

**Challenge/Flush Handling:**
- Modified challenge processing to extract resume IDs from `McpChallengeExFW` extensions
- When aggregate resume IDs are received, they are decomposed and dispatched to appropriate routes
- Added support for `McpFlushExFW` to handle tools list changed notifications with aggregate event IDs
- Flush messages now include rewritten aggregate IDs when aggregating across routes

**Configuration:**
- Updated `McpBindingConfig` to compute and store sorted prefixes and routed IDs for efficient lookup
- Added schema validation requiring `toolkit` field in route conditions for proxy bindings
- Added schema validation preventing `cache` option for server bindings

### Testing

- Added comprehensive unit tests in `McpAggregateEventIdTest` covering:
  - Prefix computation for single and multiple toolkits
  - Encoding/decoding of aggregate IDs
  - Round-trip encode/decode validation
  - Edge cases (null IDs, unknown prefixes, malformed input)
- Added integration tests verifying:
  - Tools list changed notifications with aggregate event IDs
  - Lifecycle event resumption with aggregate IDs
  - Handling of missing/unknown prefixes in aggregate IDs
- Added schema validation tests for new configuration constraints

Fixes #(issue)

https://claude.ai/code/session_018egHGhhcAWAQ1qZb6G6eva

### CODE REVIEW

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 1]
```diff
@@ -626,8 +724,34 @@ private void onClientMessage(
...
         {
-            server.doServerFlush(flush.traceId(), flush.authorization(),
-                flush.budgetId(), flush.reserved(), flush.extension());
+            final long traceId = flush.traceId();
+            final long authorization = flush.authorization();
+            final long budgetId = flush.budgetId();
+            final int reserved = flush.reserved();
+            final OctetsFW extension = flush.extension();
+
+            OctetsFW forwardExtension = extension;
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]:
```suggestion
            OctetsFW newExtension = extension;
```

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 2]
```diff
@@ -626,8 +729,34 @@ private void onClientMessage(
...
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]:
```suggestion
            OctetsFW newExtension = extension;
```

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 3]
```diff
@@ -626,8 +729,34 @@ private void onClientMessage(
...
+
+            if (server.aggregating() && extension.sizeof() > 0)
+            {
+                final McpFlushExFW flushEx =
+                    mcpFlushExRO.tryWrap(extension.buffer(), extension.offset(), extension.limit());
+                if (flushEx != null)
+                {
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]: Let's assume flushEx is non-null, so no need to check for extension.sizeof() > 0 either.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 4]
```diff
@@ -626,8 +729,34 @@ private void onClientMessage(
...
+                    final String perRouteId = extractEventId(flushEx);
+                    if (perRouteId != null)
+                    {
+                        server.recordRouteEventId(routedId, perRouteId);
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]:
```suggestion
                        server.onDecodeEventId(routedId, perRouteId);
```

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 5]
```diff
@@ -626,8 +729,34 @@ private void onClientMessage(
...
+
+            OctetsFW forwardExtension = extension;
+
+            if (server.aggregating() && extension.sizeof() > 0)
+            {
+                final McpFlushExFW flushEx =
+                    mcpFlushExRO.tryWrap(extension.buffer(), extension.offset(), extension.limit());
+                if (flushEx != null)
+                {
+                    final String perRouteId = extractEventId(flushEx);
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]:
```suggestion
                    final String eventId = extractEventId(flushEx);
```

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 6]
```diff
@@ -626,8 +729,34 @@ private void onClientMessage(
...
+                    if (perRouteId != null)
+                    {
+                        server.recordRouteEventId(routedId, perRouteId);
+                        final OctetsFW aggregateId = server.mintAggregateEventId();
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]:
```suggestion
                        final OctetsFW aggregateId = server.nextEventId();
```

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 7]
```diff
@@ -626,8 +729,34 @@ private void onClientMessage(
...
+                        if (aggregateId != null)
+                        {
+                            forwardExtension = rewriteFlushExWithAggregateId(flushEx, aggregateId);
+                        }
+                    }
+                }
+            }
+
+            server.doServerFlush(traceId, authorization, budgetId, reserved, forwardExtension);
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]: Perhaps this work to generate the aggregate event id should be done in doServerFlush instead of onClientFlush as we seems to be checking a lot of server state to decide on the flushEx to pass.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 8]
```diff
@@ -474,6 +569,7 @@ final class McpLifecycleClient
         private MessageConsumer sender;
         private int state;
         String sessionId;        // upstream-provided session id, set on BEGIN reply
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]: Why not private? Please remove comment.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 9]
```diff
@@ -563,12 +659,19 @@ private void doClientChallenge(
...
                 .wrap(codecBuffer, 0, codecBuffer.capacity())
                 .typeId(mcpTypeId)
-                .resume(b -> {})
+                .resume(b ->
+                {
+                    if (resumeId != null)
+                    {
+                        b.id(resumeId);
+                    }
+                })
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]: Perhaps it would make sense to change this to use inject with a method reference and store resumeId on the client.

File: specs/binding-mcp.spec/src/test/java/io/aklivity/zilla/specs/binding/mcp/config/SchemaTest.java [hunk 10]
```diff
@@ -39,4 +40,30 @@ public void shouldValidateServer()
...
+    public void shouldRejectServerWithCache()
+    {
+        schema.validate("server.cache.invalid.yaml");
+    }
+
+    @Test(expected = JsonValidatingException.class)
+    public void shouldRejectServerRouteWithToolkit()
+    {
+        schema.validate("server.toolkit.invalid.yaml");
+    }
```
Comment (jfallows) [inline specs/binding-mcp.spec/src/test/java/io/aklivity/zilla/specs/binding/mcp/config/SchemaTest.java:76]: Just need a positive test one each for server, proxy and client with all the supported properties set.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/config/McpBindingConfig.java [hunk 11]
```diff
@@ -57,6 +62,32 @@ public McpBindingConfig(
...
+        final String[] prefixes = routeByPrefix.keySet().stream()
+            .sorted(Comparator.naturalOrder())
+            .toArray(String[]::new);
+        final long[] routedIds = new long[prefixes.length];
+        for (int i = 0; i < prefixes.length; i++)
+        {
+            routedIds[i] = routeByPrefix.get(prefixes[i]).id;
+        }
+        this.prefixes = prefixes;
+        this.routedIds = routedIds;
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/config/McpBindingConfig.java]: Let's define a record and have a single array of records, instead of parallel arrays with same dimensions.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/config/McpAggregateEventId.java [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        {
+            return Map.of();
+        }
+
+        final List<String> distinct = new ArrayList<>(new HashSet<>(toolkits));
+        if (distinct.size() == 1)
+        {
+            return Map.of(distinct.get(0), encodeCrc32c(distinct.get(0)).substring(0, 1));
+        }
+
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/config/McpAggregateEventId.java]: We don't need this special case optimization on startup.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/config/McpAggregateEventId.java [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+        final Map<String, String> encoded = new HashMap<>();
+        for (String toolkit : distinct)
+        {
+            encoded.put(toolkit, encodeCrc32c(toolkit));
+        }
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/config/McpAggregateEventId.java]: Use `toolkits.stream().collect(toMap(k -> k, this::encodeCrc32c))`.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/config/McpAggregateEventId.java [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+        {
+            throw new IllegalStateException("unable to derive unique prefixes for toolkits: " + distinct);
+        }
+
+        final Map<String, String> result = new HashMap<>();
+        for (Map.Entry<String, String> entry : encoded.entrySet())
+        {
+            result.put(entry.getKey(), entry.getValue().substring(0, length));
+        }
+        return result;
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/config/McpAggregateEventId.java]: Use `encoded.entrySet().stream().collect(toMap(k -> k, v -> v.substring(0, length))`.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyItemFactory.java [hunk 15]
```diff
@@ -534,7 +534,7 @@ private void doClientBegin(
             lifecycle.doClientBegin(traceId);
 
             final String identifier = server.identifier;
-            final String upstreamSessionId = lifecycle.sessionId;
+            final String upstreamSessionId = lifecycle.sessionId();
             final String outboundSessionId = upstreamSessionId != null
                 ? upstreamSessionId
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyItemFactory.java]: Does it make sense to do this at the server call site instead, and pass sessionId as a parameter?

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 16]
```diff
@@ -91,6 +104,8 @@ final class McpProxyLifecycleFactory implements BindingHandler
     {
         this.writeBuffer = context.writeBuffer();
         this.codecBuffer = new UnsafeBuffer(new byte[context.writeBuffer().capacity()]);
+        this.flushExBuffer = new UnsafeBuffer(new byte[context.writeBuffer().capacity()]);
+        this.aggregateBuffer = new UnsafeBuffer(new byte[context.writeBuffer().capacity()]);
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]: This is too large, agree?

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 17]
```diff
@@ -195,6 +214,62 @@ McpLifecycleClient supplyClient(
...
+            }
+            final McpAggregateRoute[] aggregateRoutes = binding.aggregateRoutes;
+            for (int i = 0; i < aggregateRoutes.length; i++)
+            {
+                if (aggregateRoutes[i].routedId() == routedId)
+                {
+                    eventIds[i] = eventId;
+                    break;
+                }
+            }
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]: If we need lookup and store by routedId, should we store as Long2ObjectHashMap<String> instead of an array?

This has zero-allocate iterator, plus deterministic ordering.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 18]
```diff
@@ -195,6 +214,62 @@ McpLifecycleClient supplyClient(
...
+                    final McpLifecycleClient client = supplyClient(route.id);
+                    client.resumeId = eventId;
+                    if (McpState.replyOpened(client.state))
+                    {
+                        client.doClientResume(traceId, authorization);
+                    }
+                    else
+                    {
+                        client.doClientBegin(traceId);
+                    }
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]: If we make doClientBegin a no-op if initialOpening and set that state when emitting initial BEGIN, and we also make doClientResume a no-op if !replyOpened, then this becomes client.doClientBegin(...) and client.doClientResume(...) unconditionally.

Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]: Move this to a separate method to make the dispatchResume method a one-liner, then inline to dispatchResume call site.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 19]
```diff
@@ -447,10 +544,27 @@ private void doServerFlush(
...
         {
+            OctetsFW newExtension = extension;
+            if (aggregating())
+            {
+                final McpFlushExFW flushEx =
+                    mcpFlushExRO.wrap(extension.buffer(), extension.offset(), extension.limit());
+                final String eventId = extractEventId(flushEx);
+                if (eventId != null)
+                {
+                    onDecodeEventId(fromRoutedId, eventId);
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]: This part should be in onClientFlush, the rest should remain here.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 20]
```diff
@@ -246,9 +321,31 @@ private void onServerChallenge(
...
+
+            String aggregate = null;
+            if (challengeEx != null && challengeEx.kind() == McpChallengeExFW.KIND_RESUME)
+            {
+                final String16FW resumeId = challengeEx.resume().id();
+                if (resumeId != null && resumeId.length() != -1)
+                {
+                    aggregate = resumeId.asString();
+                }
+            }
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]:
```suggestion
            if (challengeEx != null &&
                challengeEx.kind() == McpChallengeExFW.KIND_RESUME)
            {
                String aggregate = challengeEx.resume().id().asString();
                ... other statements dependent on challenge resume kind
            }
```

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 21]
```diff
@@ -568,11 +688,20 @@ private void doClientResume(
             final McpChallengeExFW resumeEx = mcpChallengeExRW
                 .wrap(codecBuffer, 0, codecBuffer.capacity())
                 .typeId(mcpTypeId)
-                .resume(b -> {})
+                .resume(this::injectResumeId)
                 .build();
             doClientChallenge(traceId, authorization, resumeEx);
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]:
```suggestion
            doClientChallenge(traceId, authorization, resumeEx);
            resumeId = null;
```

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyListFactory.java
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyListFactory.java:233]: Remove comments.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyListFactory.java [hunk 22]
```diff
@@ -252,7 +252,7 @@ private void doClientBegin(
         {
             lifecycle.doClientBegin(traceId);
 
-            final String upstreamSessionId = lifecycle.sessionId;
+            final String upstreamSessionId = lifecycle.sessionId();
             final String sid = upstreamSessionId != null ? upstreamSessionId : server.lifecycle.sessionId;
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyListFactory.java]: Why not just guarantee that McpClientLifecycle stream has the correct sessionId perhaps derived from McpServerLifecycle stream, and then always use the client lifecycle sessionId here.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyItemFactory.java [hunk 23]
```diff
@@ -529,19 +529,16 @@ private McpClient(
         }
 
         private void doClientBegin(
-            long traceId)
+            long traceId,
+            String sessionId)
         {
             lifecycle.doClientBegin(traceId);
 
             final String identifier = server.identifier;
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyItemFactory.java]: Should identifier also be a parameter?

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 24]
```diff
@@ -246,9 +304,30 @@ private void onServerChallenge(
...
-                client.doClientResume(traceId, authorization);
+                final String16FW resumeId = challengeEx.resume().id();
+                final String aggregate = resumeId != null && resumeId.length() != -1
+                    ? resumeId.asString()
+                    : null;
+
+                if (aggregate != null && aggregating())
+                {
+                    McpAggregateEventId.decode(aggregate,
+                        (prefix, eventId) -> resumeClient(traceId, authorization, prefix, eventId));
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java]:
```suggestion
                        (prefix, eventId) -> onDecodeAggregateEventId(traceId, authorization, prefix, eventId));
```

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyListFactory.java
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyListFactory.java]:
```suggestion
            long originId,
            long routedId,
```

Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyListFactory.java]:
```suggestion
            this.originId = originId;
            this.routedId = routedId;
```

Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyListFactory.java]:
```suggestion
            sender = newStream(this::onClientMessage, originId, routedId, initialId,
```
Same for other usages.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java [hunk 25]
```diff
@@ -744,6 +864,70 @@ private void onClientReset(
...
+                }
+            });
+            break;
+        case McpFlushExFW.KIND_ELICIT_COMPLETE:
+            final McpElicitStatus status = flushEx.elicitComplete().status().get();
+            builder.elicitComplete(b -> b.id(buffer, offset, length).status(s -> s.set(status)));
+            break;
+        default:
+            break;
+        }
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpProxyLifecycleFactory.java:912]: Move to separate method, use inject, then inline this to call site.
Inject method is cleaner if split into separate kind-specific inject method, called from a switch statement in the generic inject.

**Review by jfallows — COMMENTED** [review]
See feedback.
